### PR TITLE
[LiveComponent] Fix loadingDirectives match elements from nested components

### LINKED ISSUE #1382: [LiveComponent] data-loading behaviour not working correctly in child components
I'm trying to add the loading state behaviour documented at https://symfony.com/bundles/ux-live-component/current/index.html#targeting-loading-for-a-specific-action to my application.

I think I'm running into a bug that is not specific for my setup (PHP 8.2.14, Symfony 5.4.33, symfony/ux-live-component 2.13.3).

I have a component that looks like this:

Class:
```php
#[AsLiveComponent(
    'test_child',
    '@test_child.component.html.twig'
)]
class TestChildComponent extends AbstractController
{
    use DefaultActionTrait;

    #[LiveAction]
    public function slowOperation(): void
    {
        sleep(2);
    }
}
```

Template:
```twig
<div {{ attributes }}>
    <a
            data-action="live#action"
            data-action-name="slowOperation"
    >
        Click me...
    </a>

    <div data-loading="action(slowOperation)|show">Loading...</div>
</div>
```

This works as expected when embedding the component into a template which is served through a "normal" Controller action:

Adobe Express 2024-01-08 09 58 01

(Please see https://www.dropbox.com/scl/fo/cifcwzq66p5fkxmw0lsh6/h?rlkey=h2o2z1753sdoqeqhis17vqf3n&dl=0 for hi-resolution versions of the screen recordings)

The "loading" message briefly flashes while the page and its component is being loaded, and is then only shown while the request for the "slowOperation" action is running, being removed when this has finished — as expected.

This behaviour becomes broken when I put the TestChildComponent into a TestParentComponent, like this:

Parent class:
```php
#[AsLiveComponent(
    'test_parent',
    '@test_parent.component.html.twig'
)]
class TestParentComponent extends AbstractController
{
    use DefaultActionTrait;
}
```

Parent template:
```twig
<div {{ attributes }}>
    <twig:test_child />
</div>
```

My assumption is that this should not change the behaviour in any way, but it does:
Adobe Express 2024-01-08 09 59 35

As before, there is only a brief flash of the loading message upon full-page-load. However, when clicking the button and thus triggering the action, the loading message is shown but *remains* shown, even after the action/request has finished.

Only subsequent clicks then result in the correct behaviour.

I'm seeing even wilder misbehaviour in my actual code, where the element with the data-loading attribute is shown (and remains shown) even when I'm only leaving a form field and no action is triggered at all.

### PR DESCRIPTION
**Pull request by smnandre**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Issues        | Fix #1382
| License       | MIT

The elements with loading directives were used in all LiveController they were in.

This PR fixes that and restrict loadingDirective to the scope of the LiveContoller (and filter those belonging to nested / child components)

### CODE REVIEW

File: src/LiveComponent/assets/dist/live_controller.js [hunk 1]
```diff
@@ -2393,7 +2393,8 @@ class LoadingPlugin {
     }
     getLoadingDirectives(element) {
         const loadingDirectives = [];
-        let matchingElements = element.querySelectorAll('[data-loading]');
+        let matchingElements = [...element.querySelectorAll('[data-loading]')];
+        matchingElements = matchingElements.filter((elt) => elt.closest('[data-controller="live"]') === element);
```
Comment (norkunas) [inline src/LiveComponent/assets/dist/live_controller.js]: This won't work when using multiple controllers, you should use `data-controller~="live"`

Comment (smnandre) [reply src/LiveComponent/assets/dist/live_controller.js]: Oh thank you! Well spotted

File: src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts [hunk 2]
```diff
@@ -150,7 +150,10 @@ export default class implements PluginInterface {
 
     getLoadingDirectives(element: HTMLElement|SVGElement) {
         const loadingDirectives: ElementLoadingDirectives[] = [];
-        let matchingElements = element.querySelectorAll('[data-loading]');
+        let matchingElements = [...element.querySelectorAll('[data-loading]')];
+
+        // ignore elements which are inside a nested "live" component
+        matchingElements = matchingElements.filter((elt) => elt.closest('[data-controller~="live"]') === element);
```
Comment (weaverryan) [inline src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts]: We have a built-in mechanism for this we should use - it's `elementBelongsToThisComponent` from `dom_utils.ts`. We can pass the `component` instance from `LoadingPlugin.attachToComponent` into the callbacks, then into here so we can call that.

**Review by weaverryan — COMMENTED** [review]
Thanks for getting on this. Possible to add a test too - should be able to combine something from `child.test.js` into `loading.test.js`.

File: src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts (hunk 2 above)
Comment (smnandre) [reply src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts]: I think using the DOM will be better in term of performance, as we just manipulate Node element.. but that could be challenged. Do you want to use it for now ?  

We could probably use only a CSS selector but i think we need to define this browser compatibility list before :)

Comment (smnandre) [reply src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts]: Test added @weaverryan

File: src/LiveComponent/assets/dist/live_controller.js [hunk 3]
```diff
@@ -2393,7 +2393,8 @@ class LoadingPlugin {
     }
     getLoadingDirectives(element) {
         const loadingDirectives = [];
-        let matchingElements = element.querySelectorAll('[data-loading]');
+        let matchingElements = [...element.querySelectorAll('[data-loading]')];
+        matchingElements = matchingElements.filter((elt) => elt.closest('[data-controller~="live"]') === element);
```
Comment (WebMamba) [inline src/LiveComponent/assets/dist/live_controller.js]: I am not sure about the use of the `closest` function. `closest` finds only one element, but in a live component you can have multiple elements with a loading attribute in nested element

File: src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts (hunk 2 above)
Comment (weaverryan) [reply src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts]: Thanks for adding the test! I still think we should use `elementBelongsToThisComponent()`. By not using it, we're adding "dangling code": we have just one part of the code that works differently than everywhere else. We would need a real, verified performance reason to do such a thing. And, except for cases where there actually *are* children (which is not most live components), the `elementBelongsToThisComponent()` is basically just a call to `element.contains()` - https://github.com/symfony/ux/blob/2.x/src/LiveComponent/assets/src/dom_utils.ts#L204

**Review by weaverryan — CHANGES_REQUESTED** [review]

File: src/LiveComponent/assets/dist/live_controller.js (hunk 3 above)
Comment (smnandre) [reply src/LiveComponent/assets/dist/live_controller.js]: closest is an ascending selector, it look the first parent "live controller"

File: src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts (hunk 2 above)
Comment (smnandre) [reply src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts]: Made changes, i cannot test locally for now, is this how you imagined it ?

File: src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts [hunk 4]
```diff
@@ -15,33 +16,33 @@ interface ElementLoadingDirectives {
 
 export default class implements PluginInterface {
     attachToComponent(component: Component): void {
-        component.on('loading.state:started', (element: HTMLElement, request: BackendRequest) => {
-            this.startLoading(element, request);
+        component.on('loading.state:started', (element: HTMLElement, request: BackendRequest, component: Component) => {
+            this.startLoading(component, element, request);
```
Comment (weaverryan) [inline src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts:20]: We already have the `component` as the argument to `attachToComponent` and we're attaching the `loading.state.started` to that specific component. So o need for the new argument to the callback: we can use the `component` variable already

**Review by weaverryan — CHANGES_REQUESTED** [review]
One minor tweak needed - thanks!

File: src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts (hunk 4 above)
Comment (smnandre) [reply src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts:20]: As i could not test, i did it in a defensive way :) 

I'll remove it

Comment (smnandre) [reply src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts:20]: Fix / rebased / rebuilded (style cannot test locally -- will fix this next week)

Comment (smnandre) [reply src/LiveComponent/assets/src/Component/plugins/LoadingPlugin.ts:20]: Think i did what you meant but not sure

**Comment by weaverryan** [comment]
Thanks Simon!

**Comment by smnandre** [comment]
Oh damn... i was pushing to rebase .... hope i did not mess on the go with things

**Comment by weaverryan** [comment]
All good, I merged before you rebased. I did the rebasing ;)

**Pull request merged**